File: src/239_Sliding_Window_Maximum/main.cpp

#include <iostream>
#include <queue>

using namespace std;

class Solution {
public:
    //Runtime: 564 ms, faster than 30.14%
    //Memory Usage: 118.7 MB, less than 52.18%
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        // 回想起这道题了，需要使用一个单调的队列，每次也要pop
        deque<int> q;
        vector<int> rtn;
        // 需要存index
        // for (int val : nums) {
        int cnt = 0;
        for (int i = 0; i < nums.size(); ++i) {
            // 想了一下，不能pop等号，要不没数据了
            // 然后另一个难点是如何保证在窗的左边界时输出最大值，即使存了index，也可能被pop
            while (q.size() && nums[q.back()] < nums[i]) {
                q.pop_back();// 留着小的数对于最大值没有帮助
            }
            // 跳出while循环表明q为空或者q里面的数比较大
            q.push_back(i);
            cnt++;
            if (cnt < k) continue;//k之前不产生数据
            
            // 滑动窗开始产生数据，此时q里面肯定是有数据的
            rtn.push_back(nums[q.front()]);
            // 在开始加入下一个数的时候pop掉过时的数据
            if (i - q.front() + 1 == k) q.pop_front();
        }
        return rtn;
    }
};

ostream & operator<< (ostream &os, vector<int> &rtn) {
    os << "[";
    for (auto val : rtn) os << val << ",";
    os << "]";
    return os;
}

int main(int argc, const char * argv[]) {
    Solution gua;
    vector<int> rtn, nums, exp;
    int k;
    rtn = gua.maxSlidingWindow(nums={1,3,-1,-3,5,3,6,7}, k = 3);
    exp = {3,3,5,5,6,7};
    cout << "expect result = " << exp << ", actual result = " << rtn << endl;

    rtn = gua.maxSlidingWindow(nums={1}, k = 1);
    exp = {1};
    cout << "expect result = " << exp << ", actual result = " << rtn << endl;

    rtn = gua.maxSlidingWindow(nums={1, -1}, k = 1);
    exp = {1, -1};
    cout << "expect result = " << exp << ", actual result = " << rtn << endl;

    rtn = gua.maxSlidingWindow(nums={4, -2}, k = 2);
    exp = {4};
    cout << "expect result = " << exp << ", actual result = " << rtn << endl;
    return 0;
}

File: src/239_Sliding_Window_Maximum/main.py

from typing import List

class Solution:
    #Runtime: 2200 ms, faster than 9.09%
    #Memory Usage: 30.5 MB, less than 17.25%
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # 这道的基本思路是单调栈
        memo, rtn = [], []
        size = len(nums)

        for i in range(size):
            # 先加入memo，加入时注意pop较小值；然后在达到滑动窗口上限时再弹出，弹出时进行相应操作

            while memo and nums[memo[-1]] < nums[i]:
                # 当然要rtn有效时才加入，相等时就不pop了
                memo.pop(-1)
            # 跳出while说明memo为空，或者前面一个数>=当前的数
            memo.append(i)

            # 以i作为标志，控制加入rtn
            if i >= k-1:
                rtn.append(nums[memo[0]])

            # 这里负责维护滑动窗的宽度限制
            if i - memo[0] + 1 == k:
                memo.pop(0) # 这次就pop吧，要不然memo里面还会剩下一个，得到for loop外面去处理

        return rtn

if __name__ == "__main__":
  gua = Solution()
  rtn = gua.maxSlidingWindow(nums=[1,3,-1,-3,5,3,6,7],k=3)
  print("expect value = [3,3,5,5,6,7], actual value = ", rtn)

  rtn = gua.maxSlidingWindow(nums=[1],k=1)
  print("expect value = [1], actual value = ", rtn)

  rtn = gua.maxSlidingWindow(nums=[1,-1],k=1)
  print("expect value = [1,-1], actual value = ", rtn)

  rtn = gua.maxSlidingWindow(nums=[9,11],k=2)
  print("expect value = [11], actual value = ", rtn)

